### PR TITLE
Implement `enumerate_adapters` for rs-backend

### LINKED ISSUE #409: Add convenience function `enumerate_adapters`
Hi wgpu-py Team,

I can not find `enumerate_adapters()` to just get an idea of all gpu available in a machine. 

is there an alternative way to do that and this is something to add?

Thank you for the feedback.

### PR DESCRIPTION
**Pull request by dskkato**
I originally used wgpu with WebGPU, but recently I came across this library. As mentioned in #409, I also needed to choose an adapter to try out compute shaders, so I implemented it on a trial basis. I'm not sure if it conforms to your coding style, but I would appreciate it if you could review it. If you are already working on something similar, please feel free to close it as appropriate.

Resolves #409

### CODE REVIEW

File: wgpu/backends/rs.py [hunk 1]
```diff
@@ -176,6 +176,20 @@ def check_struct(struct_name, d):
 libf = SafeLibCalls(lib, error_handler)
 
 
+def enumerate_adapters():
+    """Return a list of all available adapters."""
+    # The first call is to get the number of adapters, and
+    # the second call is to get the actual adapters
```
Comment (almarklein) [inline wgpu/backends/rs.py:182]: Let's mention the option here.
```suggestion
    # The first call is to get the number of adapters, and
    # the second call is to get the actual adapters.
    # Note that the second arg (now NULL) can be a
    # `WGPUInstanceEnumerateAdapterOptions` to filter by backend.
```

File: wgpu/backends/rs.py [hunk 2]
```diff
@@ -176,6 +176,20 @@ def check_struct(struct_name, d):
...
+
+    # H: size_t f(WGPUInstance instance, WGPUInstanceEnumerateAdapterOptions const * options, WGPUAdapter * adapters)
+    adapter_count = libf.wgpuInstanceEnumerateAdapters(
+        get_wgpu_instance(), ffi.NULL, ffi.NULL
+    )
+    adapters = ffi.new("WGPUAdapter[]", adapter_count)
+    libf.wgpuInstanceEnumerateAdapters(get_wgpu_instance(), ffi.NULL, adapters)
+    return [GPUAdapter(adapter) for adapter in adapters]
```
Comment (almarklein) [inline wgpu/backends/rs.py:190]: The codegen has trouble with the refactoring of the adapter metadata. A relatively easy way to work around that is to split that part out of `request_adapter()`, into a a new `GPU._create_adapter(adapter_id)`.

Then this here becomes:
```suggestion
    gpu = GPU()
    return [gpu._create_adapter(adapter) for adapter in adapters]
```

File: wgpu/backends/rs.py [hunk 3]
```diff
@@ -261,80 +275,7 @@ def callback(status, result, message, userdata):
             error_msg = error_msg or "Could not obtain new adapter id."
             raise RuntimeError(error_msg)
```
Comment (almarklein) [inline wgpu/backends/rs.py:277]:
```suggestion
        return self._create_adapter(adapter_id)

    def _create_adapter(adapter_id):
        ... and undo the deletion below :)
```

**Review by almarklein — COMMENTED** [review]
Really nice!

The codegen sometimes makes refactorings like this a bit tricky. I added a comment for a way to work around this.

I'm not 100% sure about the API, relates to #410. But the solution you offer here - to make the user go into `wgpu.backends.rs` - seems like a pretty good alternative. We'd have to figure out how to document this function though. 

This also relates somewhat to #408. I propose that we move this PR forward and figure out the #401 stuff later.

File: wgpu/backends/rs.py (hunk 2 above)
Comment (dskkato) [reply wgpu/backends/rs.py]: When I turned _create_adapter into a member function of the GPU, a TypeError occurred in enumerate_adapters as below. Therefore, I made it a free function.

```sh
_______________________________ test_rs_enumerate_adapters ________________________________

    @mark.skipif(not can_use_wgpu_lib, reason="Needs wgpu lib")
    def test_rs_enumerate_adapters():
        # Get all available adapters
>       adapters = wgpu.backends.rs.enumerate_adapters()

tests/test_rs_basics.py:152: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

    def enumerate_adapters():
        """Return a list of all available adapters."""
        # The first call is to get the number of adapters, and
        # the second call is to get the actual adapters.
        # Note that the second arg (now NULL) can be a
        # `WGPUInstanceEnumerateAdapterOptions` to filter by backend.
    
        # H: size_t f(WGPUInstance instance, WGPUInstanceEnumerateAdapterOptions const * options, WGPUAdapter * adapters)
        adapter_count = libf.wgpuInstanceEnumerateAdapters(
            get_wgpu_instance(), ffi.NULL, ffi.NULL
        )
        adapters = ffi.new("WGPUAdapter[]", adapter_count)
        # H: size_t f(WGPUInstance instance, WGPUInstanceEnumerateAdapterOptions const * options, WGPUAdapter * adapters)
        libf.wgpuInstanceEnumerateAdapters(get_wgpu_instance(), ffi.NULL, adapters)
>       gpu = GPU()
E       TypeError: 'NoneType' object is not callable

wgpu/backends/rs.py:194: TypeError
```

**Comment by dskkato** [comment]
FYI: The example for `wgpuInstanceEnumerateAdapters` I referenced is this -> https://github.com/gfx-rs/wgpu-native/blob/trunk/examples/enumerate_adapters/main.c

**Review by almarklein — APPROVED** [review]
LGTM. This is going to be interesting merging with #408 😄   I propose we merge this into `main` and then I'll merge `main` into #408.

**Review by Korijn — APPROVED** [review]

**Comment by almarklein** [comment]
Thanks for your contribution!

File: wgpu/backends/rs.py (hunk 2 above)
Comment (almarklein) [reply wgpu/backends/rs.py]: Ah yes, sorry. This is due to a bug in the `_register_backend` decorator, that will be fixed in #408.

**Comment by almarklein** [comment]
For the record, when #408 is merged, it will be `wgpu.backends.wgpu_native.enumerate_adapters()`.

**Pull request merged**